File: participants/om/ecom-app/controllers/productController.js
let products = require('../data/products');

exports.getAllProducts = (req, res) => {
    res.json(products);
};

exports.getProductById = (req, res) => {
    const productId = (req.params.id)*1;
    const product = products.find(p => p.id === productId);

    if(!product){
        return res.status(404).json({ message: "Product not found!" });
    }

    res.json(product);
};

exports.createProduct = (req, res) => {
    const { name, price, category, stock } = req.body;

    const newProduct = {
        id: products.length + 101,
        name: name,
        price: price,
        category: category,
        stock: stock
    };

    products.push(newProduct);
    res.status(201).json(newProduct);
};

exports.updateProduct = (req, res) => {
    const productId = (req.params.id)*1;
    const product = products.find(p => p.id === productId);

    if(!product){
        return res.status(404).json({ message: "Product not found!" });
    }

    const { name, price, category, stock } = req.body;
    
    product.name = name;
    product.price = price;
    product.category = category;
    product.stock = stock;

    res.status(200).json(product);
};

exports.updatePartialProduct = (req, res) => {
    const productId = (req.params.id)*1;
    const product = products.find(p => p.id === productId);

    if(!product){
        return res.status(404).json({ message: "Product not found!" });
    }

    const { name, price, category, stock } = req.body;
    
    if (name !== undefined) product.name = name;
    if (price !== undefined) product.price = price;
    if (category !== undefined) product.category = category;
    if (stock !== undefined) product.stock = stock;

    res.status(200).json(product);
};

exports.deleteProduct = (req, res) => {
    const productId = parseInt(req.params.id, 10);
    const product = products.find(p => p.id === productId);

    if(!product){
        return res.status(404).json({ message: "Product not found!" });
    }

    let newProducts = products.filter(p => p.id !== productId);
    products = newProducts;

    res.status(200).json({ message: "Product deleted successfully!" });
};

File: FastAPI/e-commerce/app/controllers/product_controller.py
from fastapi import APIRouter, HTTPException
from app.services.product_service import get_all_products, get_product_by_id, create_product as add_product, put_product, patch_product, delete_product as remove_product
from app.models.product_model import Product, ProductPatch


router = APIRouter(
    prefix="/api/product",
    tags=["Products"]
)

# @router.get("/health")
# def health_check():
#     return {"status": "Product APIs running smoothly!"}

@router.get("/")
def get_products():
    products = get_all_products()
    return products

@router.get("/{product_id}")
def get_product_id(product_id: int):
    product = get_product_by_id(product_id)
    if not product:
        raise HTTPException(status_code=404, detail="Product not found.")
    return product

@router.post("/")
def create_product(product: Product):
    new_product = add_product(product)
    return new_product

@router.put("/{product_id}")
def update_product(product_id: int, product: Product):
    updated_product = put_product(product_id, product)
    if not updated_product:
        raise HTTPException(status_code=404, detail="Product not found")
    return {
        "status" : 200,
        "success" : True,
        "data" : updated_product,
        "message" : "Product updated successfully"
    }

@router.patch("/{product_id}")
def partial_update_product(product_id: int, product: ProductPatch):
    updated_product = patch_product(product_id, product)
    if not updated_product:
        raise HTTPException(status_code=404, detail="Product not found")
    return {
        "status" : 200,
        "success" : True,
        "data" : updated_product,
        "message" : "Patch Product done successfully"
    }

@router.delete("/{product_id}")
def delete_product(product_id: int):
    deleted = remove_product(product_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Product not found")
    return {
        "status" : 200,
        "success" : True,
        "message" : "Product deleted successfully"
    }
